pypesto/hierarchical/optimal_scaling_approach/optimal_scaling_problem.py

```diff
@@ -0,0 +1,530 @@
+            )
+
+            self.groups[group]['ub_indices'] = list(
+                range(
+                    self.groups[group]['num_datapoints']
+                    + self.groups[group]['num_categories'],
+                    self.groups[group]['num_inner_params'],
+                )
+            )
+
+            self.groups[group]['C'] = self.initialize_c(group)
+
+            self.groups[group]['W'] = self.initialize_w(group)
+
+            self.groups[group]['Wdot'] = self.initialize_w(group)
+
+    @staticmethod
+    def from_petab_amici(
+        petab_problem: petab.Problem,
+        amici_model: 'amici.Model',
+        edatas: List['amici.ExpData'],
+        method: str = None,
+    ):
+        """Construct the inner problem from the `petab_problem`."""
+        if method is None:
+            method = REDUCED
+        return optimal_scaling_inner_problem_from_petab_problem(
+            petab_problem, amici_model, edatas, method
+        )
+
+    def get_groups_for_xs(self, inner_parameter_type: str) -> List[int]:
+        """Get unique list of ``OptimalScalingParameter.group`` values."""
+        groups = [x.group for x in self.get_xs_for_type(inner_parameter_type)]
+        return list(set(groups))
+
+    def get_xs_for_group(self, group: int) -> List[OptimalScalingParameter]:
+        """Get ``OptimalScalingParameter``s that belong to the given group."""
+        return [x for x in self.xs.values() if x.group == group]
+
+    def get_free_xs_for_group(
+        self, group: int
+    ) -> List[OptimalScalingParameter]:
+        """Get ``OptimalScalingParameter``s that are free and belong to the given group."""
+        return [
+            x
+            for x in self.xs.values()
+            if x.group == group and x.estimate is True
+        ]
+
+    def get_fixed_xs_for_group(
+        self, group: int
+    ) -> List[OptimalScalingParameter]:
+        """Get ``OptimalScalingParameter``s that are fixed and belong to the given group."""
+        return [
+            x
+            for x in self.xs.values()
+            if x.group == group and x.estimate is False
+        ]
+
+    def get_cat_ub_parameters_for_group(
+        self, group: int
+    ) -> List[OptimalScalingParameter]:
+        """Get ``OptimalScalingParameter``s that are category upper boundaries and belong to the given group."""
+        return [
+            x
+            for x in self.xs.values()
+            if x.group == group and x.inner_parameter_id[:6] == 'cat_ub'
+        ]
+
+    def get_cat_lb_parameters_for_group(
+        self, group: int
+    ) -> List[OptimalScalingParameter]:
+        """Get ``OptimalScalingParameter``s that are category lower boundaries and belong to the given group."""
+        return [
+            x
+            for x in self.xs.values()
+            if x.group == group and x.inner_parameter_id[:6] == 'cat_lb'
+        ]
+
+    def initialize_c(self, group):
+        """Initialize the constraints matrix for the group."""
```

Comment:
a (colloquial/detailed in words) description of what this function does / how it initializes the matrix?